extensions/smallrye-reactive-messaging/runtime/src/main/java/io/quarkus/smallrye/reactivemessaging/runtime/devmode/DevModeSupportConnectorFactoryInterceptor.java

```diff
@@ -0,0 +1,82 @@
+package io.quarkus.smallrye.reactivemessaging.runtime.devmode;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+
+import javax.annotation.Priority;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.eclipse.microprofile.reactive.streams.operators.PublisherBuilder;
+import org.eclipse.microprofile.reactive.streams.operators.ReactiveStreams;
+import org.eclipse.microprofile.reactive.streams.operators.SubscriberBuilder;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+@Interceptor
+@DevModeSupportConnectorFactory
+@Priority(Interceptor.Priority.PLATFORM_BEFORE + 10)
+public class DevModeSupportConnectorFactoryInterceptor {
+    private static Supplier<CompletableFuture<Boolean>> onMessage;
+
+    static void register(Supplier<CompletableFuture<Boolean>> onMessage) {
+        DevModeSupportConnectorFactoryInterceptor.onMessage = onMessage;
+    }
+
+    @AroundInvoke
+    public Object intercept(InvocationContext ctx) throws Exception {
+        if (onMessage == null) {
+            return ctx.proceed();
+        }
+
+        if (ctx.getMethod().getName().equals("getPublisherBuilder")) {
+            PublisherBuilder<Message<?>> result = (PublisherBuilder<Message<?>>) ctx.proceed();
+            return result.flatMapCompletionStage(msg -> {
+                CompletableFuture<Message<?>> future = new CompletableFuture<>();
+                onMessage.get().whenComplete((restarted, error) -> {
+                    if (!restarted) {
+                        // if restarted, a new stream is already running,
+                        // no point in emitting an event to the old stream
+                        future.complete(msg);
+                    }
+                });
+                return future;
+            });
+        }
+
+        if (ctx.getMethod().getName().equals("getSubscriberBuilder")) {
+            SubscriberBuilder<Message<?>, Void> result = (SubscriberBuilder<Message<?>, Void>) ctx.proceed();
+            return ReactiveStreams.fromSubscriber(new Subscriber<Message<?>>() {
+                private Subscriber<Message<?>> subscriber;
+
+                @Override
+                public void onSubscribe(Subscription s) {
+                    subscriber = result.build();
+                    subscriber.onSubscribe(s);
+                }
+
+                @Override
+                public void onNext(Message<?> o) {
+                    subscriber.onNext(o);
+                    onMessage.get().join();
```

Comment:
Won't this block the IO thread? Also won't the restart happen after the message has been delivered, rather than before, so the message is handled by the old code?

Reply:
Hum, right - this method is going to be called on this I/O thread. Why is it require to join actually? 

In theory, the connectors are passing the reactive streams TCK meaning that the onNext / onError / onComplete methods cannot be called concurrently (but could be called by different threads). 

Also, we may want to reverse the `onNext` and `onMessage` call to handle the message with the new code.

Reply:
I actually did try something like `onMessage.get().whenComplete((value, error) -> subscriber.onNext(o))`, but that only works well with instrumentation-based reload. With full restart, you've got a new stream running, and no way to push the message there.

Note that this is `@Outgoing` -- so there's no user code handling the message (unless we had some kind of "message interceptors" per https://github.com/smallrye/smallrye-reactive-messaging/issues/1141 :-) ). It's only the connector code at this point, submitting the message to a broker (or whatever is on the other side).

I trigger restart _after_ submitting the message, so that the message isn't lost. Note that even if I did triggered restart _before_ submitting the message, the exact same thing would happen with full restart anyway, because full restart takes a few seconds during which the old code still runs. (At least that's what I observed.) And I wait for restart to finish so that no other message may come in. I realize blocking is a crude way to do it, but I figured that wouldn't be such a big deal in dev mode.

I'm open to other ideas of course.

I toyed with the idea of using the request protocol to handle this, but I really don't feel like I could do it correctly :-)